modern/src/Actions.ts

```diff
@@ -34,8 +34,33 @@ export function gotSkinBlob(blob: Blob, store: ModernStore) {
   };
 }
 
+async function unloadSkin(makiTree) {
+  await Utils.asyncTreeFlatMap(makiTree, async (node: MakiObject) => {
+    switch (node.name) {
+      case "groupdef": {
+        // removes groupdefs from consideration (only run scripts when actually referenced by group)
+        return {};
+      }
+      case "script": {
+        if (node.system) {
+          node.system.onscriptunloading();
+        }
+        return node;
+      }
+      default: {
+        return node;
+      }
+    }
+  });
+}
+
 export function gotSkinZip(zip: JSZip, store: ModernStore) {
   return async dispatch => {
+    // unload current skin if one has been loaded
+    if (store.getState().modernSkin.skinLoaded) {
+      unloadSkin(store.getState().modernSkin.makiTree);
```

Comment:
Should we dispatch something here to communicate that the skin is unloaded?

Reply:
Yea, sounds good